File: content/snippets/python/interview_lcci_02_06_palindrome_linked_list.py
class ListNode:
    def __init__(self, val: int, next: "ListNode | None" = None) -> None:
        self.val = val
        self.next = next


def build_list(nums: list[int]) -> ListNode | None:
    dummy = ListNode(0)
    tail = dummy
    for n in nums:
        tail.next = ListNode(n)
        tail = tail.next
    return dummy.next


def reverse_list(head: ListNode | None) -> ListNode | None:
    prev = None
    while head is not None:
        nxt = head.next
        head.next = prev
        prev = head
        head = nxt
    return prev


def is_palindrome(head: ListNode | None) -> bool:
    if head is None or head.next is None:
        return True

    slow = head
    fast = head
    while fast is not None and fast.next is not None:
        slow = slow.next
        fast = fast.next.next

    second = reverse_list(slow)
    p1 = head
    p2 = second
    while p2 is not None:
        if p1 is None or p1.val != p2.val:
            return False
        p1 = p1.next
        p2 = p2.next
    return True


def main() -> None:
    a = build_list([1, 2, 2, 1])
    b = build_list([1, 2, 3, 2, 1])
    print(is_palindrome(a))
    print(is_palindrome(b))


if __name__ == "__main__":
    main()

File: content/snippets/scala2/interview_lcci_02_06_palindrome_linked_list.scala
object Lcci0206PalindromeLinkedList {
  final case class ListNode(var value: Int, var next: ListNode | Null = null)

  def buildList(nums: Vector[Int]): ListNode | Null = {
    val dummy = ListNode(0, null)
    var tail: ListNode = dummy
    nums.foreach { n =>
      tail.next = ListNode(n, null)
      tail = tail.next.asInstanceOf[ListNode]
    }
    dummy.next
  }

  def reverseList(head: ListNode | Null): ListNode | Null = {
    var prev: ListNode | Null = null
    var cur = head
    while (cur != null) {
      val node = cur.asInstanceOf[ListNode]
      val nextNode = node.next
      node.next = prev
      prev = node
      cur = nextNode
    }
    prev
  }

  def isPalindrome(head: ListNode | Null): Boolean = {
    if (head == null || head.asInstanceOf[ListNode].next == null) return true

    var slow = head
    var fast = head
    while (fast != null && fast.asInstanceOf[ListNode].next != null) {
      slow = slow.asInstanceOf[ListNode].next
      fast = fast.asInstanceOf[ListNode].next.asInstanceOf[ListNode].next
    }

    var p1 = head
    var p2 = reverseList(slow)
    while (p2 != null) {
      if (p1.asInstanceOf[ListNode].value != p2.asInstanceOf[ListNode].value) return false
      p1 = p1.asInstanceOf[ListNode].next
      p2 = p2.asInstanceOf[ListNode].next
    }
    true
  }

  def main(args: Array[String]): Unit = {
    val a = buildList(Vector(1, 2, 2, 1))
    val b = buildList(Vector(1, 2, 3, 2, 1))
    println(isPalindrome(a))
    println(isPalindrome(b))
  }
}
